services.yaml を基底とするようにロード順序を修正

nanasess opened the pull request:
<!-- 以下を参考にコメントを作成してください。 -->

## 概要(Overview・Refs Issue)
[支払方法が異なる商品を同時にカートに入れられるようにする](http://doc4.ec-cube.net/customize_service#支払方法が異なる商品を同時にカートに入れられるようにする)カスタマイズ例が動作しない問題の対策

packages 以下の yaml よりも services.yaml の方が優先順位が高い状態となっていた

## 方針(Policy)
services.yaml は以下の順序でロードされるように修正

1. app/config/eccube/services.yaml
2. app/config/eccube/packages/<package_name>.yaml
3. app/config/eccube/packages/<APP_ENV>/<package_name>.yaml
4. app/config/eccube/services_<APP_ENV>.yaml
5. app/Plugin/<PluginCode>/Resource/config/services.yaml

## 参考

https://github.com/EC-CUBE/ec-cube/issues/3959#issuecomment-434986331

## テスト（Test)
TODO

## 相談（Discussion）
app/config/eccube/packages/<package_name>.yaml で、今まで動いてなかったものが無いか確認が必要

## マイナーバージョン互換性保持のための制限事項チェックリスト
<!-- マイナーバージョンでは、機能・プラグイン・デザインテンプレート互換性を損なう変更は原則取り込みません。 -->
<!-- 以下の変更を行っていないことを確認してください。変更を行っていなければチェックしてください。 -->

- [x] 既存機能の仕様変更
- [x] フックポイントの呼び出しタイミングの変更
- [x] フックポイントのパラメータの削除・データ型の変更
- [x] twigファイルに渡しているパラメータの削除・データ型の変更
- [x] Serviceクラスの公開関数の、引数の削除・データ型の変更
- [x] 入出力ファイル(CSVなど)のフォーマット変更

nanasess committed:
services.yaml を基底とするようにロード順序を修正

okazy commented:
## ロード順について補足

- 後に読み込まれた設定が有効となる（上書きされる）
- `services.yaml` でデフォルトの設定をしている
```yaml
services:
    _defaults:
        autowire: true
        autoconfigure: true
        public: false
```
- `services.yaml` 内の `autowire:true` で基本的なDIは自動で設定される
- 後勝ちとなるため、 `packages/<package_name>.yaml` 等の設定が `services.yaml` で上書きされてしまう可能性がある状態となっていた
- 設定の優先度が変わるので、現状の機能に影響がないか確認する必要がある

chihiro-adachi commented:
以下2点を確認し、影響範囲を見極めましょう

- packages/*以下で定義されているものが動作しているか(修正前は定義しても動かないものがあった)
- services.yamlで定義したものがpackages/*の定義で上書かれるようになるので、こちらも確認が必要

nanasess commented:
twig_extension.yaml に `_defaults` の記述があるが、これは不要と思われる。

https://github.com/EC-CUBE/ec-cube/blob/5e64b926d3e91ec6b3b70457acc83574e0e032ca/app/config/eccube/packages/twig_extensions.yaml#L2-L5

nanasess commented:
- packages/mobile_detect.yaml は何も設定されていないため不要
- packages/dev/monolog.yaml と services_dev.yaml の設定が重複している(services_dev.yaml の方を削除)

nanasess committed:
不要な services の設定を削除

- packages/mobile_detect.yaml は何も設定されていないため削除
- packages/dev/monolog.yaml と services_dev.yaml の設定が重複しているため削除

chihiro-adachi commented:
ドキュメントのカスタマイズ例でカート分割〜購入できることを確認しました。
